infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/NamespaceNameValidator.java

```diff
@@ -51,6 +53,18 @@ public static boolean isValid(String name) {
     return validateInternal(name).isOk();
   }
 
+  public static String normalize(String namespaceName) {
```

Comment:
what's wrong with:

```java
StringBuilder sb = new StringBuilder(namespaceName);

// remove invalid characters at the start of the name
while (!sb.isEmpty()) {
  if (!isValidNamespaceNameStartOrEnd(sb.charAt(0))) {
    sb.remove(0);
  }
}

// corner case handling
if (sb.isEmpty()) throw AreYouKiddingMeException();
if (sb.length() == 1) return sb.toString();

// replace characters in the middle of the name with '-'.
int idx = 1;
do {
  if (!isValidNamespaceNameChar(sb.charAt(idx))) {
    sb.replace(idx, idx+1, "-");
  }
} while (++idx < sb.length());

// remove invalid characters from the end of the name
while (!sb.isEmpty()) {
  if (!isValidNamespaceNameStartOrEnd(sb.charAt(sb.length() - 1))) {
    sb.remove(sb.length() - 1);
  }
}

// corner case handling
if (sb.isEmpty()) throw new AreYouKiddingMeException();

// done
return sb.toString();
```

```java
boolean isValidNamespaceNameStartOrEnd(char c) {
   return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9');
}

boolean isValidNamespaceNameChar(char c) {
  return c == '-' || isValidNamespaceNameStartOrEnd(c);
}
```

The above is just as quick and IMHO more comprehensible than inverted regexes, etc.